fabtests/component/sock_test.c

```diff
@@ -0,0 +1,253 @@
+#include <sys/time.h>
+#include <netdb.h>
+#include <time.h>
+#include <poll.h>
+#include <sys/epoll.h>
+
+#define PORT 8888
+
+size_t server_addr_len;
+
+int get_server_addr(char* server_addr, struct sockaddr_storage *server_sock, int *sock)
+{
+	struct addrinfo *res;
+	struct sockaddr_in *in;
+	struct sockaddr_in6 *in6;
+	int ret;
+	int optval = 1;
+
+	ret = getaddrinfo(server_addr, NULL, NULL, &res);
+	if (ret) {
+		printf("getaddrinfo failed\n");
+		return ret;
+	}
+
+	memcpy(server_sock, res->ai_addr, res->ai_addrlen);
+	server_addr_len = res->ai_addrlen;
+
+	switch (server_sock->ss_family) {
+	case AF_INET:
+		in = (struct sockaddr_in *) server_sock;
+		in->sin_port = PORT;
+		break;
+	case AF_INET6:
+		in6 = (struct sockaddr_in6 *) server_sock;
+		in6->sin6_port = PORT;
+		break;
+	default:
+		printf("Unsupported Address family\n");
+		ret = -1;
+		break;
+	}
+
+	*sock = socket(server_sock->ss_family, SOCK_STREAM, 0);
+	if (sock < 0)
+		return -1;
+
+	ret = setsockopt(*sock, SOL_SOCKET, SO_REUSEADDR, (char *) &optval, sizeof(optval));
+	if (ret) {
+		printf("error setting socket options\n");
+		return ret;
+	}
+
+	ret = bind(*sock, (struct sockaddr *)server_sock, server_addr_len);
+
+	return ret;
+}
+
+int server_create(int connections, struct sockaddr_storage server_sock, int sock, int** clients)
+{
+	int new_sock, ret, i;
+
+	ret = listen(sock, connections);
+	if (ret) {
+		printf ("listen failed\n");
+		return ret;
+	}
+
+	*clients = calloc(connections, sizeof(int));
+	if (!*clients)
+		return -1;
+
+	for (i = 0; i < connections; i++) {
+		new_sock = accept(sock, NULL, NULL);
+		if (new_sock < 0) {
+			printf("error during server init\n");
+			return -1;
+		}
+
+		(*clients)[i] = new_sock;
+		
+	}
+	return 0;
+}
+
+int client_connect(int connections, struct sockaddr_storage server_addr, int sock, int **clients)
+{
+	int i, ret, new_sock;
+	int optval = 1;
+
+	*clients = calloc(connections, sizeof(int));
+
+
+	for (i = 0; i < connections; i++) {
+
+		new_sock = socket(server_addr.ss_family, SOCK_STREAM, 0);
+		if (new_sock < 0)
+			return -1;
+
+		ret = setsockopt(new_sock, SOL_SOCKET, SO_REUSEADDR, (char *) &optval, sizeof(optval));
+		if (ret) {
+			printf("error setting socket options\n");
+			return ret;
+		}
+
+		ret = connect(new_sock, (struct sockaddr *)&server_addr, server_addr_len);
+		if (ret)
+			return ret;
+
+		(*clients)[i] = new_sock;
+	}
+
+	return ret;
+}
+
+double time_select(int connections, int server_sock, int *clients) 
+{
+	fd_set select_set;
+	int i, max_sock;
+	struct timeval timeout;
+	struct timespec start;
+	struct timespec stop;
+
+	timeout.tv_sec = 0;
+	timeout.tv_usec = 0;
+
+	FD_ZERO(&select_set);
+	FD_SET(server_sock, &select_set);
+
+	max_sock = server_sock;
+
+	for (i = 0; i < connections; i++) {
+		FD_SET(clients[i], &select_set);
+
+		if (clients[i] > max_sock)
+			max_sock = clients[i];
+
+	}
+
+	clock_gettime(CLOCK_MONOTONIC, &start);
+	select(max_sock + 1, &select_set, NULL, NULL, &timeout);

```

Comment:
The timeout would immediately expire because it's set to 0 (`sec = usec = 0`). I don't know the exact intention, but I would personally expect `select()` to block indefinitely (e.g., the `timeout` argument would be NULL).

Reply:
The code looks out of date.  The intent to is time how long it takes to call select when X number of sockets are being monitored.  This should be done by timing a loop where select is called a few thousand times, then calculating an average.  We want to see if select is faster than poll than epoll, and where the crossover points are based on how many sockets we want to monitor.